Refactor, Protocol Enhancements, and Device Ecosystem Reorganization

MicBoucinha opened the pull request:
> [!IMPORTANT]
> We started working on this PR by merging the contributions made in [PR#1](https://github.com/harp-tech/pyharp/pull/1)


## Refactoring & Project Structure
- Refactored enums, class members, and function naming for consistency with official Harp naming conventions
- Changed namespace to `harp` and sub-namespace reorganization:
    - `harp.communication` for communication-related modules
    - `harp.protocol` for core protocol definitions
    - `harp.devices` for device-specific implementations (using [PEP420](https://peps.python.org/pep-0420/) for implicit namespace packages)

## Features & Enhancements
- Updated Harp protocol support with missing features
    - Added support for device serial number and missing registers
    - Improved message handling (arrays, timestamps, error detection) for different payloads
- Updated currently known Harp devices names
- Added context manager support to `Device`
- Added static factory method `HarpMessage.create`
- Changed dump file convention to be at the `Device` level instead of message level

- Removed `ReadXXXHarpMessage` and `WriteXXXHarpMessage` classes and `ReadXXX` and `WriteXXX` static methods from `HarpMessage` and moved them to `Device`
- Removed existing `behavior` driver because of new `harp.behavior` namespace package

## Documentation
- Added MkDocs documentation for API, and project structure, and per-device documentation
- Use `numpy` style convention for docstrings documentation
- Added docstrings to all public methods and classes
- Moved and updated existing examples to `docs` folder
- Created GitHub Action to generate documentation (manually triggered)

Documentation here: https://fchampalimaud.github.io/pyharp

## Testing
- Updated some tests as needed

## Technical 
- Convert project to `uv`

## Contributors
- Major contributions by Luís Teixeira and José Grilo

## harp.devices
- Automatically generated python package for each device and respective documentation
- Added `harp.devices` namespace package for device-specific implementations

Repository here: https://github.com/fchampalimaud/harp.devices

---- 


As mentioned in [issue #2](https://github.com/harp-tech/pyharp/issues/2), we were really interested in developing the Python Harp ecosystem further.

Our starting point was [PR #1](https://github.com/harp-tech/pyharp/pull/1), which already had pushed `pyharp` forward (e.g., by implementing a way to deal with the device's events), and try to build on top of that. At first, we mainly focused on completing what was missing regarding the protocol (e.g., add a generic way to deal with the different payload types) and the interaction with a generic Harp device (e.g., add methods for the missing common registers).

Our goal was to have a simple enough but versatile `pyharp` interface and we based our rationale to be as similar as possible as the Bonsai.Harp implementation. One way to achieve that was to get inspiration from the `CreateMessage` Bonsai node to implement the `HarpMessage.create` static method. Additionally, we added methods for the `Device` class that allow the users to read/write from/to registers that have different payload types.

Another thing we tried to focus on was documentation. It surely can be improved but, at the moment, we documented every function and class of the project in order to facilitate the API usage and adapted/added some minimal examples that should help the user get started with the `pyharp` package.

Beyond being possible to connect and interact with a generic Harp device, we also wanted to generate device-specific interfaces. On the one hand, we would like that each Harp device had its own package, so that if, for example, a user just wants to use `Harp Behavior`, it doesn't need to import a package that contains the interfaces for every board. This would also help when there's individual updates on each device. On the other hand, we would like that every package (including `pyharp` "core") had the same feeling, in the sense the user would get the implicit idea that every package is part of the same ecosystem. In order to achieve that, we organized the project directories according to [PEP420](https://peps.python.org/pep-0420/), in which the global ecosystem has the following structure:

```
harp/
├── communication/
├── data (harp-python content hopefully)
├── protocol/
└── devices/
    ├── behavior/
    ├── device_a/
    ├── ...
    └── device_z/
```

We also wanted to have a way to generate the device-specific packages automatically, so that we could easily add new devices and keep the documentation up-to-date. For that, we used [reflex-generator](https://github.com/harp-tech/reflex-generator) to generate the device-specific packages based on the device's YAML. Currently, the generated packages are available in the `harp.devices` repository mentioned above, but we believe that the Python interface should reside in each device's repository. [PR#86](https://github.com/harp-tech/reflex-generator/pull/86) to reflex-generator has been submitted to support the Python package generation.

Finally, we abandoned the `pyharp` name for the package because, unfortunately, someone registered a package in PyPI with the same name. The package name is currently `harp-protocol`. At the same time, and mostly to keep coherence with the namespace naming convention used on C#/Bonsai side, we decided to change the global namespace to `harp`.


## Known limitations of the PEP420 usage
To allow the usage of PEP420, we cannot have members under `harp` or `harp.devices` that are not namespace packages. This means that we cannot have a `__init__.py` file under those directories. As a consequence, we cannot have any code directly in those paths, which means that we cannot have any imports there.

This also means that the `harp-python` package cannot currently be installed in the same venv until some changes happen there. Our suggestion at the time was to place it's contents under `harp.data` and then import it from there. This would allow the `harp-python` package to be installed in the same venv as `harp-protocol` and the individual harp devices. We also submitted a PR to the `harp-python` repository to reflect this change in [PR#52](https://github.com/harp-tech/harp-python/pull/52).

With these new PRs, we aim to expand the current features of the Python implementation and make it more accessible for new users, helping them adopt and engage with the HARP ecosystem. We also see value in consolidating the Python implementation within the official harp-tech repository, and we’re more than happy to help maintain it. This would help avoid fragmented development efforts and lead to a more unified and robust solution for the community. We will continue iterating on these developments and commit new features to this PR while we work toward a consensus on merging it into the harp-tech repository.

Poofjunior committed:
can now pip instal in editable mode

Poofjunior committed:
updating readme

Poofjunior committed:
fixing readme bug

Poofjunior committed:
adding udev rules

Poofjunior committed:
device_names is now a default dict; examples now also work on linux

Poofjunior committed:
fixing pdf link; tweaking serial port to use ftdi rule

Poofjunior committed:
Update README.md

Poofjunior committed:
PEP 563 return type hints

Poofjunior committed:
grand messages.py refactor

Poofjunior committed:
more message.py refactor

Poofjunior committed:
adding harp rules

Poofjunior committed:
adding simple behavior io stuff

Poofjunior committed:
adding prelim behavior driver

Poofjunior committed:
adding a string representation

Poofjunior committed:
added __str__ and active mode script

Poofjunior committed:
printing DI event changes works

Poofjunior committed:
updating how payload is parsed

Poofjunior committed:
adding enabling of events

Poofjunior committed:
adding inWaiting to reads

Poofjunior committed:
read dumped registers

Poofjunior committed:
expose more core features

Poofjunior committed:
Merge pull request #1 from AllenNeuralDynamics/dev/editable_mode

Dev/editable mode

Poofjunior committed:
add WriteU32 and WriteS32

Poofjunior committed:
fix signed error in WriteS32

Poofjunior committed:
add debug level logging.

jessy-liao committed:
Added ReadS32 and refactored WriteS32 to handle array inputs

Poofjunior committed:
Merge pull request #2 from jessy-liao/master

Added ReadS32 and refactored WriteS32 to handle array inputs

patricklatimer committed:
Add threaded serial port

patricklatimer committed:
clean up formatting changes

patricklatimer committed:
add event_count method and update test

patricklatimer committed:
remove unused import

patricklatimer committed:
undo formatting change

patricklatimer committed:
remove python-versions pin

patricklatimer committed:
add future import for list annotations

patricklatimer committed:
fix future import

Poofjunior committed:
Merge pull request #4 from AllenNeuralDynamics/feat/threading

Add threaded serial port

Poofjunior committed:
update wait-for-events example

Poofjunior committed:
remove extraneous call

Poofjunior committed:
Merge pull request #5 from AllenNeuralDynamics/feat/example_update

update wait-for-events example

MicBoucinha committed:
Convert project to uv

MicBoucinha committed:
Define docstyle convention

ZegCricket committed:
Update default vscode settings

ZegCricket committed:
Add docs to the project

ZegCricket committed:
Delete poetry.lock

ZegCricket committed:
Update .gitignore and README

MicBoucinha committed:
Merge pull request #1 from fchampalimaud/feature/docs_init

Add documentation to the project

ZegCricket committed:
Chore: remove shebangs from example scripts

MicBoucinha committed:
Cleanup unused code

MicBoucinha committed:
Remove ReadXXHarpMessage classes

MicBoucinha committed:
Move base classes

MicBoucinha committed:
Formatting

MicBoucinha committed:
Update protocol link

MicBoucinha committed:
Convert usage of ReadXX classes to device related methods

MicBoucinha committed:
Remove ReadXX methods from HarpMessage

MicBoucinha committed:
Extract unrelated Enum members

MicBoucinha committed:
Add read_S64 and read_U64 methods to Device

MicBoucinha committed:
Create write_XXX methods on Device

MicBoucinha committed:
Merge branch 'feature/refactor_messages' into develop

MicBoucinha committed:
Add dump parameter to Device's write methods

MicBoucinha committed:
Change default name and changed visibility of control registers methods

MicBoucinha committed:
Refactor timeout assignment and correct payload string formatting

MicBoucinha committed:
Update current_device_names

MicBoucinha committed:
Add support for device's serial number

MicBoucinha committed:
Remove unused method

MicBoucinha committed:
Update README

MicBoucinha committed:
Update documentation dependencies

MicBoucinha committed:
Update mkdocs configuration

MicBoucinha committed:
Move examples to documentation folder

MicBoucinha committed:
Update documentation

MicBoucinha committed:
Add context manager support to Device class

ZegCricket committed:
Chore: add pytest configs to vscode workspace settings

ZegCricket committed:
Docs: separate examples in different pages

ZegCricket committed:
Docs: modify documentation organizational, functional and visual aspects

ZegCricket committed:
Docs: modify home page

ZegCricket committed:
Test: comment incorrect tests to fix in the future

ZegCricket committed:
Docs: add documentation to HarpMessage and Device classes

ZegCricket committed:
Refactor: change base.py enums to IntEnums

ZegCricket committed:
Refactor: rename functions to follow a consistent naming pattern

ZegCricket committed:
Refactor: move DeviceMode to base.py

ZegCricket committed:
Refactor: make some class members private

ZegCricket committed:
Refactor: modify way of sending write harp messages

ZegCricket committed:
Refactor: misc changes

ZegCricket committed:
Feature: add HarpMessage.create static method

ZegCricket committed:
Docs: add comments to the remaining code

ZegCricket committed:
Feature: add classes in base.py to root pyharp namespace

ZegCricket committed:
Feature: remove read_xx and write_xx Device methods

ZegCricket committed:
Docs: reorganize examples section

ZegCricket committed:
Docs: add olfactometer example

ZegCricket committed:
Docs: add pyharp core API to documentation

MicBoucinha committed:
Merge branch 'feature/update_docs' into develop

MicBoucinha committed:
Change dump file usage convention

MicBoucinha committed:
Change DeviceMode to OperationMode for official Harp documentation coherence

MicBoucinha committed:
Change return type of Device's send method

MicBoucinha committed:
Update and add missing OperationControl methods

MicBoucinha committed:
Update docs links

MicBoucinha committed:
Fix doc link

MicBoucinha committed:
Comment device tests

MicBoucinha committed:
Add pytest-cov dev dependency

MicBoucinha committed:
Add is_error to HarpMessage

MicBoucinha committed:
Read Timestamp entry for PayloadType

MicBoucinha committed:
Handle HarpMessage with arrays and timestamps

Also moved the common properties to HarpMessage

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Restore read and write methods from Device

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Remove dump from send method

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add ClockConfig and TimestampOffset

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add OperationCtrl and ResetMode

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Allow Port control

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Fix issue with payload property

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update payload_as_string

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update CommonRegisters members name

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update send method to receive HarpMessage

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Remove unused methods

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Fix bug with uninitialized fields and wrong property call

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Fix bug on Device's send method

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Handle empty replies

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Fix bug with str method

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Reformat

MicBoucinha committed:
Update pre-commit config

MicBoucinha committed:
Add monorepo-plugin to mkdocs

MicBoucinha committed:
Fix typo

MicBoucinha committed:
Test new module structure

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update project

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Merge branch 'feature/pyharp_namespace' into develop

MicBoucinha committed:
Delete behavior driver

MicBoucinha committed:
Move device.py and harp_serial.py to communication folder

MicBoucinha committed:
Move core from docs to protocol

MicBoucinha committed:
Update docs links with correct namespaces

MicBoucinha committed:
Remove duplicated extension from mkdocs config

MicBoucinha committed:
Minor fix on example

MicBoucinha committed:
Update files to prevent circular imports

MicBoucinha committed:
Update project related information for pushing it to PyPI

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Change namespace to "harp"

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update minimum Python version

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add "include markdown plugin" to docs

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update version

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add dependency to handle dataclasses on documentation generation

MicBoucinha committed:
Update mkdocs config

- Removed emojis
- Added footer next/prev navigation
- Added copyright to footer
- Added github link to footer

MicBoucinha committed:
Update docs

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add CurrentDriver to known device names

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Reorder docs' TOC

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Remove type annotation on Enum

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update version

MicBoucinha committed:
Add github action to build and deploy docs

MicBoucinha committed:
Update LICENSE

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

glopesdev commented:
@MicBoucinha @bruno-f-cruz @ZegCricket summary of proposal:

- Deprecate `harp-python`
- Deploy three packages:
  - `harp-protocol`: contains a common (ideally Pydantic compatible) model for common protocol types, namespace `harp.protocol`
  - `harp-serial`: contains the implementation of the serial port `Device` in Python, namespace `harp.serial`
  - `harp-data`: contains what is currently in `harp-python`, but refactored to pick types from `harp.protocol`, namespace would be `harp.data`
- Namespace packages:
  - `harp`: core root folder for all other packages
  - `harp.devices`: common root folder for all auto-generated pyharp interface packages
- Work on moving `pyharp` repo to `harp-python` and refactor `harp-python` as a multi-project repo

Decide on whether to use `.` or `-` as package separator.

glopesdev commented:
We should keep issues like these in mind: https://github.com/pypa/pip/issues/3666.

Indeed it seems like there is a normalization scheme in place, but also for backwards compatibility reasons it might be best to avoid `.` as a separator, and possibly converge on `-` but we can keep discussing.

Btw, the below works:
<img height="100" alt="image" src="https://github.com/user-attachments/assets/0d040c50-4033-4954-812c-d30acaed5f42" />

Respectively, `uv pip install harp.protocol` also works, so it seems both are equivalent, but for package listing it might be best to have a single consistent naming convention.

ZegCricket commented:
I personally like the idea of using `.` as a separator in order to distinguish the `harp-tech` Python packages from the remaining non-`harp-tech` Harp packages. At the same time, it would have the same feel as the `Bonsai`/`dotnet` package IDs and some of the repo names used in the `harp-tech` organization (such as `core.pico` or `device.behavior`).

With that said, I don't mind using whatever convention it's decided in the end.

glopesdev commented:
> I personally like the idea of using . as a separator in order to distinguish the harp-tech Python packages from the remaining non-harp-tech Harp packages.

@ZegCricket which other harp-tech Harp packages do we have? Why is it important to distinguish them from other harp packages in this way? Are they not normal Python packages?

Edit: nevermind, I misunderstood your comment. I still think using `.` as a way to "tag" or "reserve" a package prefix is still not ideal, as the distinction only shows up in pypi, and I don't feel people search the PyPI index that much (I feel usually people they at the package install instructions by other means, e.g. website, and it works either way).

MicBoucinha committed:
Rename communication namespace to 'serial'

MicBoucinha committed:
Remove unnecessary devices folder

MicBoucinha committed:
Fix issue with Visual_en flag

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Fix documentation path for Device

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Implement a Timeout handling strategy

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update payload handling taking into account the timeout strategy in send method

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Reorganize project directory structure

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Create workspaces for harp-protocol and harp-serial

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update docs

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

ZegCricket committed:
Chore: add pyrefly extension to `extensions.json`

ZegCricket committed:
Chore: add script to run docs locally on Windows

ZegCricket committed:
Add harp reference epoch constant

ZegCricket committed:
Fix issues with types

ZegCricket committed:
Docs: fix imports from examples to take the new project structure into account

ZegCricket committed:
Refactor: simplify WriteHarpMessage handling of PayloadType

ZegCricket committed:
Chore: add VSCode setting that enables pyrefly to display type errors

MicBoucinha committed:
Removed custom message from exceptions

MicBoucinha committed:
Pin harp-protocol version on harp-serial

MicBoucinha committed:
Fix: raise HarpReadException for missing timestamp in ReplyHarpMessage

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add LaserDriverController to device_names

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Refactor payload handling in HarpMessage to properly support Python 3.9

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add operation_ctrl methods to device

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update documentation

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Fix tests

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Merge branch 'feature/workspaces' into develop

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/harp-tech/pyharp?pullRequest=3) <br/>All committers have signed the CLA.

MicBoucinha commented:
Hi! We made the following changes since the last meeting:

- Separate the previous `harp-protocol` into 2 different packages (`harp-protocol` and `harp-serial`), using the `uv` workspaces feature
- Implement a strategy to handle timeouts when waiting for a reply from the device (`harp.serial.TimeoutStrategy`)

## Notes

1. `harp-protocol` & `harp-serial` are already released as well as new versions of [harp.devices](https://github.com/fchampalimaud/harp.devices) using `harp-serial` directly.
  
2. [Documentation](https://fchampalimaud.github.io/pyharp) is online and updated.
  
  @glopesdev @bruno-f-cruz, feel free to give us feedback where improvements can be done.

bruno-f-cruz commented on .github/workflows/build-docs.yml:
```diff
@@ -0,0 +1,56 @@
+      - name: Clone harp.devices repository
+        uses: actions/checkout@v4
+        with:
+          repository: fchampalimaud/harp.devices
```
I am not sure the package repository should be coupled with the generation of docs for the devices.

MicBoucinha replied:
In the long term, we could consider to separate the docs generation from the packages in different repos. The current solution is not set in stone, but we would like to keep the documentation for the generic device and for the specific devices together.

bruno-f-cruz commented on .gitignore:
.vscode should be added to the list

MicBoucinha replied:
The reason we keep the `.vscode` directory in the repo is because it has the extensions (and respective settings) we actively use for development and that we would like for contributors to use as well. Do you have an alternative suggestion regarding this matter?

bruno-f-cruz replied:
I would not include it at all since vscode is not required for development and even if people use vscode they may use different extensions. Anything that you want contributors to follow should be enforced via standard metadata files and linting (e.g. pyproject.toml and github actions)

bruno-f-cruz commented on LICENSE:
This should be changed to harp-tech, similar to what happens with the other software packages

MicBoucinha replied:
We can change it when the move of the repo to `harp-tech` is imminent.

bruno-f-cruz commented on 10-harp.rules:
Not sure what this config file is for? But should likely be ignored?

MicBoucinha replied:
This file was inherited from the AIND fork, but we agree it can be removed.

bruno-f-cruz commented on LICENSE:
```diff
@@ -1,6 +1,7 @@
 MIT License
 
 Copyright (c) 2020 OEPS & Filipe Carvalho
```
We should consider making this standard with the copyright in the other software packages.

https://github.com/harp-tech/harp-python/blob/3d97d0b800533e4821896f744e52569c6413c6f7/LICENSE#L3

MicBoucinha replied:
Sure, same as answered previously, 

> We can change it when the move of the repo to harp-tech is imminent.

bruno-f-cruz commented on docs/examples/olfactometer_example/olfactometer_example.md:
Same question as above. Is it really necessary to reference specific instances of devices in this generic package?
If the goal is to show how to interface with a device, why not simply use registers in the core and consider the device an abstract instance?

bruno-f-cruz commented on docs/examples/read_and_write_from_registers/read_and_write_from_registers.md:
Same as above. Consider using write registers in the core instead of specific devices

MicBoucinha replied:
We are currently writing new examples (based on the ones that exist on the [CF.Bonsai](https://fchampalimaud.github.io/cf.bonsai/) website) and rewriting the ones that currently exist. As mentioned above, we would like to keep the documentation for the whole `pyharp` ecosystem in the same website, so we would like to have examples for a generic device and for the specific ones together as well.

bruno-f-cruz commented on mkdocs.yml:
```diff
@@ -0,0 +1,96 @@
+site_name: pyharp
+repo_url: "https://github.com/fchampalimaud/pyharp"
```
Must be updated

MicBoucinha replied:
Will update it when the documentation is generated from `harp-tech`.

bruno-f-cruz commented on mkdocs.yml:
```diff
@@ -0,0 +1,96 @@
+site_name: pyharp
+repo_url: "https://github.com/fchampalimaud/pyharp"
+# repo_name: "pyharp"
+copyright: Copyright &copy; 2025, Hardware and Software Platform, Champalimaud Foundation
```
Should be harp-tech and collaborators

MicBoucinha replied:
Sure, same as answered previously,

> We can change it when the move of the repo to harp-tech is imminent.

bruno-f-cruz commented on mkdocs.yml:
```diff
@@ -0,0 +1,96 @@
+            show_source: false
+            extensions:
+              - griffe_fieldz
+  - git-committers:
```
Unless we are maintaining this automatically, you should consider removing

MicBoucinha replied:
We didn't understand this comment, can you please elaborate?

bruno-f-cruz replied:
I thought you had to list each contribuitor individually, but it seems the extension uses the repo metadata to do it for you? Ignore this if that is indeed the case.

bruno-f-cruz commented on mkdocs.yml:
```diff
@@ -0,0 +1,96 @@
+        icon: material/weather-night
+        name: Switch to system preference
+
+nav:
```
This is not required for a merge, but we should probably have a pre-build step on the mkdocs build process that automatically generates the nav from the existing folder structure (API and Examples at least).  I have hacked something together in the past (https://github.com/AllenNeuralDynamics/contraqctor/blob/main/docs/_regenerate_api.py) but we can def do better.

MicBoucinha replied:
We can discuss it further to find a better alternative to what is currently implemented.

bruno-f-cruz commented on pyproject.toml:
```diff
@@ -1,23 +1,57 @@
-python = "^3.4"
-pyserial = "^3.4"
+[project.urls]
+Repository = "https://github.com/fchampalimaud/pyharp/"
```
must be updated

MicBoucinha replied:
Sure, same as answered previously,

> We can change it when the move of the repo to harp-tech is imminent.

bruno-f-cruz commented on pyproject.toml:
```diff
@@ -1,23 +1,57 @@
-[tool.poetry.scripts]
-pyharp = "pyharp.main:main"
+[dependency-groups]
+dev = [
```
This should be split into two groups: dev and docs. No need to bootstrap docs for testing

MicBoucinha replied:
Sure, we can do that

bruno-f-cruz commented on pyproject.toml:
```diff
@@ -1,23 +1,57 @@
+]
+
+[tool.ruff.lint.pydocstyle]
+convention = "numpy"
```
This is probably a petpeeve of mine, but can we use google over numpy? It reads must cleaner.

MicBoucinha replied:
This can be considered in the future, but it's not a priority right now. 

From our current usage pattern (usually through VSCode), either one shows up nicely on the editor.

bruno-f-cruz commented on run_docs.ps1:
I think the doc build scripts can be completely removed if we dont try to generate docs for all the device instances. Am I correct? This would greatly simply the action as we would only need to call mkdocs build.

Alternatively, consider using mkdocs hooks to generate tihs instead of shell scripts.

MicBoucinha replied:
That specific script is for local development on Windows. We agree that we could eventually handle this through mkdocs hooks, but we didn't really explore that option yet. We can look into it.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/device_names.py:
```diff
@@ -0,0 +1,52 @@
+# commit used: https://github.com/harp-tech/protocol/commit/3e2a228
+
+current_device_names = {
+    256: "USBHub",
```
Same point as before. I dont think this should exist. If it does, then I think it should be automatically generated via a submodule reference

MicBoucinha replied:
We agree, we need to find a better way to do this. We can discuss this further.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+REFERENCE_EPOCH = datetime(1904, 1, 1)
+
+# Bit masks for the PayloadType
+_isUnsigned: int = 0x00
```
This should be camel case

MicBoucinha replied:
I think you meant snake case, but agreed.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+from datetime import datetime
+from enum import IntEnum, IntFlag
+
+# The reference epoch for UTC harp time
+REFERENCE_EPOCH = datetime(1904, 1, 1)
+
+# Bit masks for the PayloadType
+_isUnsigned: int = 0x00
+_isSigned: int = 0x80
+_isFloat: int = 0x40
+_hasTimestamp: int = 0x10
+
+
+class MessageType(IntEnum):
+    """
+    An enumeration of the allowed message types of a Harp message. More information on the MessageType byte of a Harp message can be found [here](https://harp-tech.org/protocol/BinaryProtocol-8bit.html#messagetype-1-byte).
+
+    Attributes
+    ----------
+    READ : int
+        The value that corresponds to a Read Harp message (1)
+    WRITE : int
+        The value that corresponds to a Write Harp message (2)
+    EVENT : int
+        The value that corresponds to an Event Harp message (3). Messages of this type are only meant to be send by the device
+    READ_ERROR : int
+        The value that corresponds to a Read Error Harp message (9). Messages of this type are only meant to be send by the device
+    WRITE_ERROR : int
+        The value that corresponds to a Write Error Harp message (10). Messages of this type are only meant to be send by the device
+    """
+
+    READ = 1
+    WRITE = 2
+    EVENT = 3
+    READ_ERROR = 9
+    WRITE_ERROR = 10
+
```
```suggestion
class MessageType(IntFlag):
    READ = 0x01
    WRITE = 0x02
    EVENT = 0x03
    ERROR = 0x08
    READ_ERROR = READ | ERROR
    WRITE_ERROR = WRITE | ERROR
```

Consider using a IntFlag instead. 
It is kinda weird that:
print(Foo.READ & Foo.EVENT) evaluates to foo.READ, but it is also the case using the current Enum anyway.

MicBoucinha replied:
We will look into your suggestion.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+    S8 = _isSigned | 1
+    U16 = _isUnsigned | 2
+    S16 = _isSigned | 2
+    U32 = _isUnsigned | 4
```
Use hex for consistency

MicBoucinha replied:
Certainly.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+    Float = _isFloat | 4
+    Timestamp = _hasTimestamp
+    TimestampedU8 = _hasTimestamp | U8
+    TimestampedS8 = _hasTimestamp | S8
```
Following pep8 (https://docs.python.org/3/howto/enum.html) this should be TIMESTAMPED_S8

MicBoucinha replied:
Certainly.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+    HW_VERSION_H = 0x01
+    HW_VERSION_L = 0x02
+    ASSEMBLY_VERSION = 0x03
+    HARP_VERSION_H = 0x04
```
This should read CORE_VERSION not HARP_VERSION

MicBoucinha replied:
It was inherited, but sure.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+from datetime import datetime
+from enum import IntEnum, IntFlag
+
```
I think we need an internal variable of the library to track the version of the protocol this is targeting.

MicBoucinha replied:
Do you have any specific suggestions?

bruno-f-cruz replied:
Maybe a simple:

```python
__PROTOCOL_VERSION__ = 1.15
```

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+        Bit 7 (0x80): If set to 1, the device sends an Event Message with the R_TIMESTAMP_SECONDS content each second (heartbeat)
+    """
+
+    OP_MODE = 3 << 0
+    DUMP = 1 << 3
+    MUTE_RPL = 1 << 4
+    VISUALEN = 1 << 5
+    OPLEDEN = 1 << 6
+    ALIVE_EN = 1 << 7
```
Use hex for consistency

MicBoucinha replied:
Sure

bruno-f-cruz commented on src/harp-protocol/harp/protocol/exceptions.py:
```diff
@@ -0,0 +1,32 @@
+    Exception raised when there is an error writing to a register in the Harp device.
+    """
+
+    def __init__(self, register):
```
This should be type annotated

MicBoucinha replied:
Agreed.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/exceptions.py:
```diff
@@ -0,0 +1,32 @@
+
+    def __init__(self, timeout_s: float):
+        super().__init__(f"No reply received within {timeout_s} seconds.")
+        self.timeout_s = timeout_s
```
I wonder if this should have some sort of context of the message that made the request that is being awaited.

MicBoucinha replied:
It makes sense, we will look into it. Same for the other exceptions.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/messages.py:
```diff
@@ -0,0 +1,582 @@
+                ]
+
+        elif pt == PayloadType.S16 or pt == PayloadType.TimestampedS16:
+            if self.length == payload_start + 2:
```
This sort of predicates cna be simplified but only checking against the existance of the S16 flag.

MicBoucinha replied:
We understand that but we decided to follow the Zen of Python:

> Explicit is better than implicit.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/messages.py:
```diff
@@ -0,0 +1,582 @@
+                ]
+
+        elif pt == PayloadType.S8 or pt == PayloadType.TimestampedS8:
+            if self.length == payload_start + 1:
```
I may be missing something, but it looks like this can be made much simpler by using the offset of the datatype (1, 2, 4, etc...)

MicBoucinha replied:
Same as before, we prefer to be more explicit.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/messages.py:
```diff
@@ -0,0 +1,582 @@
+        else:
+            raise Exception(
+                "The value cannot be None if the message type is equal to MessageType.WRITE!"
+            )
```
Not sure why this is the case. Why isnt it valid to build other kinds of messages?

MicBoucinha replied:
In principle, the boards should be the only ones to send event messages, but this may be rethinked: we may allow the creation of a broader set of messages and only restrain in specific use cases (like when interacting with a device).

bruno-f-cruz commented on src/harp-protocol/harp/protocol/messages.py:
```diff
@@ -0,0 +1,582 @@
+    _PAYLOAD_CONFIG = {
+        # payload_type: (byte_size, signed, is_float)
+        PayloadType.U8: (1, False),
+        PayloadType.S8: (1, True),
```
This is what you should use in the base class to parse the payloads.

MicBoucinha replied:
We will have to look at this with more detail.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/messages.py:
```diff
@@ -0,0 +1,582 @@
+        )
+
+
+class ReplyHarpMessage(HarpMessage):
```
I am a bit confused by this API choice. What do you gain by using specific classes instead of always using HarpMessage as the input/output of the device? There is some syntactic sugar, but if you want it I would suggest turning the constructors into HarpMessage.create_write_message(...) instead of using inheritance. This way we always maintain the same class for transactions.

MicBoucinha replied:
This was inherited and we just followed from there. There are some specific methods for each class, but we can reevaluate the situation.

bruno-f-cruz commented on src/harp-protocol/pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+keywords = ['python', 'harp']
+requires-python = ">=3.9,<4.0"
+
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
```
Any reason to prefer hatchling over the default uv build (https://docs.astral.sh/uv/reference/settings/#build-backend)?

MicBoucinha replied:
This was the default when we started the project and so far has been working fine, but we can change this if it's relatively easy to adapt.

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1485 @@
+from __future__ import annotations  # enable subscriptable type hints for lists.
+
```
Not sure what this comment refers too?

MicBoucinha replied:
The comment can go, it was inherited.

bruno-f-cruz commented on src/harp-protocol/harp/protocol/base.py:
```diff
@@ -0,0 +1,280 @@
+    TimestampedFloat = _hasTimestamp | Float
+
+
+class CommonRegisters(IntEnum):
```
The name should be singular

MicBoucinha replied:
We tried to follow a similar convention that was used in Bonsai.Harp (see here: https://github.com/bonsai-rx/harp/blob/35b69c64c1fcf359aefbf884008303d0d5a9bbd6/src/Bonsai.Harp/DeviceRegisters.cs).

bruno-f-cruz replied:
Uhm the issue is that in the C# API its a static class whereas here is a Enum. If it is an Enum, it should be singular, especially in this case where only one flag can be picked at a time.

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1485 @@
+        The serial number of the device
+    """
+
+    WHO_AM_I: int
```
Not a strong opinion, however, as you already have the common registers defined in the protocol. I think this would be a bit easier to maintain by having a method:
Device.CoreRegisters[CommonRegister.WHO_AM_I].

MicBoucinha replied:
From what we understood from what you said, we think it makes the code more confusing. I think we need to discuss it further.

bruno-f-cruz replied:
My worry is that you are representing the same thing in two different places. Seems super error prone if we ever update.

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1485 @@
+        self,
+        serial_port: str,
+        dump_file_path: Optional[str] = None,
+        read_timeout_s: float = 1,
```
Simplify the name to: `timeout`

MicBoucinha replied:
Agreed

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1485 @@
+            The binary file to which all Harp messages will be written
+        read_timeout_s: float, optional
+            _TODO_
+        """
```
This is not up to date with the signature

MicBoucinha replied:
Thank you! We will update it.

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1485 @@
+            rtscts=True,
+        )
+
+        # open file if it is defined
+        if self._dump_file_path is not None:
+            self._dump_file = open(self._dump_file_path, "ab")
```
Not sure I follow how you would want to use this.
I can see cases where users may want to skip querying the device for metadata, but this suggests that a valid use-case is to cache it?

MicBoucinha replied:
We have to look at this more carefully. We can discuss it further.

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1485 @@
+        reply = self.send(HarpMessage.create(MessageType.READ, address, PayloadType.U8))
+        return OperationMode(reply.payload & OperationCtrl.OP_MODE)
+
+    def dump_registers(self) -> list:
```
I think we should discuss a bit this API synchronously during one of the meetings. I dont dislike this but it has the potential to grow a bit out of control.

MicBoucinha replied:
This was only for this specific register where we don't know yet how to constrain the user's input regarding the OP_MODE to be of type OperationCtrl. In terms of usage, for this particular register, having the separated method made more sense. But we should discuss it further.

bruno-f-cruz commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,22 @@
+repos:
```
I would suggest turning this into an action instead of a pre-commit.

MicBoucinha replied:
Why an action only?
I might be missing something, but personally, I prefer to have checks locally before committing when possible.
We can discuss this further.

bruno-f-cruz replied:
Pre-commits are something we need to rely on people installing. GitHub actions are something we can enforce during PRs. 
Pre-commits can be bypassed locally, github actions can't
I can't verify the output of a pre-commit hook that runs in your PC, but I can verify the log of a GitHub action.

You are right that doesnt need to be an action "only", but if we are maintaining one thing, let it be the github action.

bruno-f-cruz reviewed (changes requested):
Thanks for taking a stab at this. I left some first pass feedback. 
I think we should discuss a bit the architecture of the device API synchronously during one of the meetings. I dont have a specific issue with it, but I am just afraid it may become hard to maintain if each register has its own specific read/write method with a name given without any sort of consistency.

MicBoucinha committed:
Remove unnecessary udev rules file

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Refactor HARP_VERSION to CORE_VERSION

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update documentation

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Refactor base.py enums for consistency and clarity

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Refactor HarpMessage to consolidate message creation

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Remove HarpMessage.create static method

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Refactor HarpMessage to remove ReplyHarpMessage

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Refactor timeout variables

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Rename HarpTimeoutError to HarpTimeoutException and add context to it

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update comment

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Add HarpMessage context to Harp exceptions

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Remove constraint when creating HarpMessage

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Change build system

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Remove device_names

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Handle potential serial connection errors

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update minimum Python version

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

ZegCricket committed:
Fix: write events to dump file

MicBoucinha committed:
Merge remote-tracking branch 'origin/feature/dump_events_to_file' into feature/pr_review_changes

MicBoucinha committed:
Refactor to improve message reading and timeout handling in Device

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Separate docs from dev dependencies

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Bump version of packages

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Update documentation

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>

MicBoucinha committed:
Merge branch 'feature/pr_review_changes' into develop

bruno-f-cruz commented on .gitignore:
```diff
@@ -1,6 +1,174 @@
+__pypackages__/
+
+# Celery stuff
+celerybeat-schedule
```
I am wondering where this .gitignore is coming from. It seems quite opinionated. Is there a reason to exclude so much stuff? Are you really expecting people to use pdm, celery and flask to develop this library?

bruno-f-cruz commented on src/harp-protocol/harp/protocol/messages.py:
```diff
@@ -0,0 +1,471 @@
+    DEFAULT_PORT: int = 255
+    BASE_LENGTH: int = 4
+    _frame: bytearray = bytearray()
+    _port: int = DEFAULT_PORT
```
Consider removing the port setter and the `DEFAULT_PORT` constant, and simply passing it as an argument, with a default, in the constructor. Also consider changing to Hex notation to match what is written in the [protocol (0xFF)](https://github.com/harp-tech/protocol/blob/main/BinaryProtocol-8bit.md#payloadtype-1-byte).

bruno-f-cruz commented on src/harp-serial/README.md:
`````diff
@@ -0,0 +1,75 @@
+
+Then reload udev rules with
+````
+sudo udevadm control --reload-rules
`````
Can you explain why this is required?

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+from __future__ import annotations  # for type hints (PEP 563)
```
Remove comment

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+    TIMESTAMP_OFFSET: int
+
+    _ser: HarpSerial
+    _dump_file_path: Optional[Path]
```
Is this actually being used anywhere?

I like the idea of having a logger that makes everything log in the standard format. But this should be decoupled from the device class.
Maybe lets remove this implementation for now?

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+
+        return reply is not None
+
+    def reset_device(
```
We should consider removing these syntactic sugar methods and just create message constructors for the core registers instead. Something closer to what we have in the csharp api.

In other words:
```
reset_message = ResetMessage(...)
device.send(reset_message)

```

Seems a bit more scalable in cases where the protocol changes.

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+        if self._dump_file:
+            self._dump_file.write(reply)
+
+    def get_events(self) -> list[HarpMessage]:
```
This is grabing not only events but also other messages in the queue, no?

bruno-f-cruz reviewed (changes requested):
A few more notes to go over during the SRM meeting

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+    def load(self) -> None:
+        """
+        Loads the data stored in the device's common registers.
+        """
```
```python

class CoreRegisters # This should be generated from the core.yml in theory
   WHOAMI = RegConfiguration

for reg in CoreRegisters:
   device.read(reg)
```

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+        if self._dump_file_path is not None:
+            self._dump_file = open(self._dump_file_path, "ab")
+
+    def disconnect(self) -> None:
```
refactor to match serial and call it `close()`

bruno-f-cruz commented on src/harp-serial/harp/serial/device.py:
```diff
@@ -0,0 +1,1347 @@
+
+        return reply
+
+    def send(
```
```python 
# Device knows how to read and write Register types
class Device:
   def write(Register): # Register can be core or application specific
   def read(Register):

  def request/send/command(HarpMessage)

class HarpBehavior(Device):
   ...

# Registers have a reg and know how to parse/compose payloads from value
class RegisterPayload:
  _reg = 32
  def make_payload(...)

# The API thus looks like:
behavior.write(DigitalOutputSetPayload(value)) -> DigitalOutpuSetPayload
behavior.write(RegisterPayload<T>) -> RegisterPayload<T>
```

MicBoucinha committed:
Update docs

Co-authored-by: José Grilo <jose.grilo@research.fchampalimaud.org>
